plugin/pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -637,82 +647,79 @@ func NewServiceAffinityPredicate(podLister algorithm.PodLister, serviceLister al
 }
 
-// CheckServiceAffinity ensures that only the nodes that match the specified labels are considered for scheduling.
-// The set of labels to be considered are provided to the struct (ServiceAffinity).
-// The pod is checked for the labels and any missing labels are then checked in the node
-// that hosts the service pods (peers) for the given pod.
+func (s *ServiceAffinity) UpdatePredicateMetaServicePods(selector labels.Selector, predicateMeta *predicateMetadata) error {
+	var err error
+	if predicateMeta.servicePods == nil || time.Since(predicateMeta.servicePodsLastRefreshTime) > PREDICATE_METADATA_REFRESH {
+		predicateMeta.lock.Lock()
+		defer predicateMeta.lock.Unlock()
+		predicateMeta.servicePods, err = s.podLister.List(selector)
+		predicateMeta.servicePodsLastRefreshTime = time.Now()
+	}
+	return err
+}
+
+// The CheckServiceAffinity predicate matches nodes in such a way to force that
+// ServiceAffinity.labels are homogenous for pods added to a node.
 //
-// We add an implicit selector requiring some particular value V for label L to a pod, if:
-// - L is listed in the ServiceAffinity object that is passed into the function
-// - the pod does not have any NodeSelector for L
-// - some other pod from the same service is already scheduled onto a node that has value V for label L
+// (i.e. it returns true IFF this pod can be added to this node, such
+// that all other pods in the same service are running on nodes w/
+// the exact same ServiceAffinity.label values).
+//
+// Details:
+// First check if the pod provides enough constraints to gaurantee all replicas
+// will have homogenous affinity labels.
+//
+// If not, then we need to create an "implicit selector" which constraints anything not already constrained
+// by the pod itself, this is done by introspecting against pods that are already in the cluster.
+//
+// If (the affinity labels are a SUBSET of the pod's constraints)
+// 	The pod has all information necessary to check affinity, the pods label selector is sufficient to calculate
+// 	the match.
+// Otherwise:
+// 	Create an "implicit selector" which gaurantees homogenous labels for all pods, like so.
+// 	Missing labels in the pod: Fill them in by picking a node on the cluster, which already fulfills this svc,
+//	and use that pods labelling information as to fill in any remaining affinity constraints.
+// 	To do this, we "reverse engineer" a selector by introspecting existing pods running under the same service/namespace.
+// 		- L is a label that the ServiceAffinity object needs as a matching constraints.
+// 		- L is not defined in the pod.
+// 		- AND SOME pod from a service is the same namespace, already scheduled onto a node provides that value.
 func (s *ServiceAffinity) CheckServiceAffinity(pod *api.Pod, meta interface{}, nodeInfo *schedulercache.NodeInfo) (bool, []algorithm.PredicateFailureReason, error) {
+	if _, ok := meta.(*predicateMetadata); !ok {
+		panic("Failed casting predicate metadata")
```

Comment:
Can you really please use the construct from here:
https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/algorithm/predicates/predicates.go#L458

?